frontend/src/util/composables.ts

```diff
@@ -14,56 +15,68 @@ import {
   useUrlSearchParams,
 } from "@vueuse/core";
 
-// reactive variable synced with url params, as object of strings
-const params = useUrlSearchParams("history"); // only supports replace, not push
+/** reactive variable synced with url params, as object of strings. only supports replace, not push */
+const params = useUrlSearchParams("history");
 
-// generic param type
+/** generic param type */
 type Param<T> = {
   parse: (value: string) => T;
   stringify: (value: T) => string;
 };
 
-// param treated as string
+/** param treated as string */
 export const stringParam: Param<string> = {
   parse: (value) => value,
   stringify: (value) => String(value),
 };
 
-// param treated as number
+/** param treated as number */
 export const numberParam: Param<number> = {
   parse: (value) => Number(value) || 0,
-  // 5 decimals good enough for lat/long:
-  // https://en.wikipedia.org/wiki/Decimal_degrees#Precision
+  /** a few decimals good enough for lat/long: */
+  /** https://en.wikipedia.org/wiki/Decimal_degrees#Precision */
   stringify: (value) => String(round(value || 0, 5)),
 };
 
-// param treated as boolean
+/** param treated as boolean */
 export const booleanParam: Param<boolean> = {
   parse: (value) => (value.toLowerCase() === "true" ? true : false),
   stringify: (value) => String(value),
 };
 
-// reactive variable synced with a specific url param
-// no good third party solution exists for this, so write our own basic version
-// see https://github.com/vueuse/vueuse/issues/3398
+/** param treated as array of other params */
+export const arrayParam = <T>(param: Param<T>): Param<T[]> => ({
+  parse: (value) => value.split(",").map(param.parse),
+  stringify: (value) => value.map(param.stringify).join(","),
+});
+
+/** reactive variable synced with a specific url param */
+/** no good third party solution exists for this, so write our own basic version */
+/** see https://github.com/vueuse/vueuse/issues/3398 */
 export function useUrlParam<T>(
   name: string,
   { parse, stringify }: Param<T>,
   initialValue: T,
 ) {
-  // https://github.com/vuejs/composition-api/issues/483
+  /** https://github.com/vuejs/composition-api/issues/483 */
   const variable = shallowRef(initialValue);
 
-  // when url changes, update variable
-  watchEffect(() => {
-    const param = params[name] || "";
-    const value = parse(Array.isArray(param) ? param.join() : param);
-    if (value) variable.value = value;
-  });
-
-  // when variable changes, update url
+  /** when url changes, update variable */
+  watch(
+    () => params[name],
+    () => {
+      const param = params[name] || "";
+      const value = parse(Array.isArray(param) ? param.join() : param);
+      if (param === stringify(variable.value)) return;
```

Comment:
Check if param is already equal to value, to avoid unnecessary state updates. Also incorporate stringify, because stringifying can lose information, e.g. numberParam rounds to 5 decimals.

Reply:
I thought I had added a comment for this but maybe it got deleted in the shuffle. I needed to do this because the stringifying process is sometimes "lossy". For example, in the case of geo coordinates, they get rounded to the 5th decimal place. So the incoming `variable.value` might be `0.123456789`, but the param in the url is `0.12346`, and we want to consider those as equal. I'll add a comment.